tests/0104-fetch_from_follower_mock.c

```diff
@@ -0,0 +1,171 @@
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "test.h"
+
+
+/**
+ * @name Fetch from follower tests using the mock broker.
+ */
+
+
+/**
+ * @brief Test offset reset when fetching from replica.
+ */
+static void do_test_offset_reset (const char *auto_offset_reset) {
+        const char *bootstraps;
+        rd_kafka_mock_cluster_t *mcluster;
+        rd_kafka_conf_t *conf;
+        rd_kafka_t *c;
+        const char *topic = "test";
+        const int msgcnt = 1000;
+        const size_t msgsize = 1000;
+
+        TEST_SAY(_C_MAG "[ Test FFF auto.offset.reset=%s ]\n",
+                 auto_offset_reset);
+
+        mcluster = test_mock_cluster_new(3, &bootstraps);
+
+        /* Seed the topic with messages */
+        test_produce_msgs_easy_v(topic, 0, 0, 0, msgcnt, msgsize,
+                                 "bootstrap.servers", bootstraps,
+                                 "batch.num.messages", "10",
+                                 NULL);
+
+        /* Set partition leader to broker 1, follower to broker 2 */
+        rd_kafka_mock_partition_set_leader(mcluster, topic, 0, 1);
+        rd_kafka_mock_partition_set_follower(mcluster, topic, 0, 2);
+
+        test_conf_init(&conf, NULL, 0);
+        test_conf_set(conf, "bootstrap.servers", bootstraps);
+        test_conf_set(conf, "auto.offset.reset", auto_offset_reset);
+        /* Make sure we don't consume the entire partition in one Fetch */
+        test_conf_set(conf, "fetch.message.max.bytes", "100");
+
+        c = test_create_consumer("mygroup", NULL, conf, NULL);
+
+        /* The first fetch will go to the leader which will redirect
+         * the consumer to the follower, the second and sub-sequent fetches
+         * will go to the follower. We want the third fetch, second one on
+         * the follower, to fail and trigger an offset reset. */
```

Comment:
I think you might mean to use `RD_KAFKA_RESP_ERR_OFFSET_NOT_AVAILABLE` here?

`RD_KAFKA_RESP_ERR_OFFSET_OUT_OF_RANGE` shouldn't trigger an offset reset unless `fetch_offset` is less than `HighwaterMarkOffset` (which would seem the least important case of the two to test, but we could have tests for both). It should bounce back to the leader, which then decides what to do next - there could be different test cases for different scenarios, which don't seem to be defined in this test either.

Reply:
I added an additional test for lagging replicas (trigger ERR_OFFSET_NOT_AVAILABLE)